File: go/tree/98_is_valid_bst.go
package tree

import "math"

// isValidBST 是否是一个有效的二叉搜索树
// 二叉搜索树 左子树都比根节点小，右子树都比根节点大
// DFS
func isValidBST(root *TreeNode) bool {

	var helper func(*TreeNode, int, int) bool

	helper = func(node *TreeNode, lower, upper int) bool {
		if node == nil {
			return true
		}

		// lower, upper 代表左子树和右子树的值
		// BST 根节点应该大于左子树的值，小于右子树的值
		if node.Val <= lower || node.Val >= upper {
			return false
		}

		return helper(node.Left, lower, node.Val) && helper(node.Right, node.Val, upper)
	}

	return helper(root, math.MinInt64, math.MaxInt64)
}

// 中序遍历
func isValidBST2(root *TreeNode) bool {

	if root == nil {
		return true
	}

	// 中序遍历，迭代实现
	stack := make([]*TreeNode, 0)

	// ans := make([]int, 0)
	prev := math.MinInt64

	node := root

	for len(stack) > 0 || node != nil {

		// 先遍历左子树
		for node != nil {
			stack = append(stack, node)
			node = node.Left
		}
		// 根节点
		popped := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		// ans = append(ans, popped.Val)
		// if len(ans) > 1 && ans[len(ans)-1] <= ans[len(ans)-2] {
		// 	return false
		// }
		// 优化上面的ans ，不用O(n) 的空间
		if popped.Val <= prev {
			return false
		}
		prev = popped.Val
		// if popped.Val <= node.Val {
		// 	return false
		// }
		// 右子树
		node = popped.Right
	}
	return true
}

File: rust/src/tree/98_is_valid_bst.rs



// is_valid_bst 验证二叉树, 中序遍历
pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {

    if root.is_none() {
        return true
    }
    // 中序遍历，迭代实现
    let mut stack = vec![];
    let mut node = root;
    let mut prev = i64::MIN;

    while !stack.is_empty() || node.is_some() {

        // 先遍历左子树
        while let Some(n) = node {
            node = n.borrow_mut().left.take();
            stack.push(n);
        }

        // 根节点，BST 的特征 根节点 >= 左子树，<= 右子树，否则不是BST
        if let Some(popped) = stack.pop(){
            if popped.borrow().val as i64 <= prev {
                return false;
            }
            prev = popped.borrow().val as i64;
            node = popped.borrow_mut().right.take();
        }
    }
    true
}


// is_valid_bst 验证二叉树， DFS
pub fn is_valid_bst2(root: Option<Rc<RefCell<TreeNode>>>) -> bool {

    helper(&root, i64::MIN, i64::MAX)

}

fn helper(root: &Option<Rc<RefCell<TreeNode>>>, lower: i64, upper: i64) -> bool {

    if root.is_none() {
        return true
    }

    // BST 的性质，根节点大于左子树，小于右子树
    if let Some(n) = root {
        if n.borrow().val as i64 <= lower || n.borrow().val as i64 >= upper {
            return false
        }

        let v = n.borrow().val as i64;
        return helper(&n.borrow().left, lower, v)
        && helper(&n.borrow().right, v, upper) 
    }
    
    true
}